### ISSUE TITLE
Ray 1.2.0 fails to connect to Ray cluster on K8s (running master)

### ISSUE THREAD

**Issue by ssiegel95** [ray-project/ray#14799]
Working through the instructions given on https://docs.ray.io/en/master/cluster/kubernetes.html, I'm running into the following issue on this step:

```
Then open a new shell and try out a sample program:

$ python ray/doc/kubernetes/example_scripts/run_local_example.py
The program in this example uses ray.util.connect(127.0.0.1:10001) to connect to the Ray cluster.
```
```bash
Traceback (most recent call last):
  File "ray/doc/kubernetes/example_scripts/run_local_example.py", line 57, in <module>
    ray.util.connect(f"127.0.0.1:{LOCAL_PORT}")
  File "/home/stus/miniconda3/envs/fctk/lib/python3.7/site-packages/ray/util/client_connect.py", line 26, in connect
    conn_str, secure=secure, metadata=metadata, connection_retries=3)
  File "/home/stus/miniconda3/envs/fctk/lib/python3.7/site-packages/ray/util/client/__init__.py", line 57, in connect
    connection_retries=connection_retries)
  File "/home/stus/miniconda3/envs/fctk/lib/python3.7/site-packages/ray/util/client/worker.py", line 120, in __init__
    raise ConnectionError("ray client connection timeout")
ConnectionError: ray client connection timeout
```

I've confirmed that the service is available:

```
$  microk8s.kubectl -n ray get services
NAME                       TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                       AGE
example-cluster-ray-head   ClusterIP   10.152.183.99   <none>        10001/TCP,8265/TCP,8000/TCP   16m
```

And I've enabled port forwarding:

```
$  microk8s.kubectl -n ray port-forward service/example-cluster-ray-head 10001:10001
Forwarding from 127.0.0.1:10001 -> 10001
Forwarding from [::1]:10001 -> 10001
Handling connection for 10001
Handling connection for 10001
```

I ran this all from the current master (cd89f0dc55ae98231aa08e9a0e1c80409e75acf1).

Any help would be appreciated. Thanks.

**Comment by MarcoAlejandro** [comment]
I'm having the same problem. 

Kubernetes Ray cluster:
```
$ kubectl -n ray get pods
NAME                         READY   STATUS    RESTARTS   AGE
ray-head-7749d67596-s246j    1/1     Running   0          29m
ray-worker-697869545-cf29c   1/1     Running   0          29m
ray-worker-697869545-q6csw   1/1     Running   0          28m
ray-worker-697869545-snktp   1/1     Running   1          29m
```

Cluster services:
```
kubectl -n ray get services
NAME       TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                       AGE
ray-head   ClusterIP   10.245.231.51   <none>        10001/TCP,8265/TCP,6379/TCP   60m
```

And the port `10001` is being forwarded. 

Exception message: 

```
Traceback (most recent call last):
  File "cluster_exp.py", line 55, in <module>
    ray.util.connect(f"127.0.0.1:{LOCAL_PORT}")
  File "/Users/marcosandoval/Library/Caches/pypoetry/virtualenvs/ray-P2KvvS3W-py3.8/lib/python3.8/site-packages/ray/util/client_connect.py", line 25, in connect
    return ray.connect(
  File "/Users/marcosandoval/Library/Caches/pypoetry/virtualenvs/ray-P2KvvS3W-py3.8/lib/python3.8/site-packages/ray/util/client/__init__.py", line 53, in connect
    self.client_worker = Worker(
  File "/Users/marcosandoval/Library/Caches/pypoetry/virtualenvs/ray-P2KvvS3W-py3.8/lib/python3.8/site-packages/ray/util/client/worker.py", line 120, in __init__
    raise ConnectionError("ray client connection timeout")
ConnectionError: ray client connection timeout
```

**Comment by MarcoAlejandro** [comment]
Documentation for `v1.2.0` has a different page. I will follow that version. 

https://docs.ray.io/en/releases-1.2.0/cluster/kubernetes.html#ray-k8s-run

**Comment by ssiegel95** [comment]
Thanks @MarcoAlejandro I will give this a try.

**Comment by ssiegel95** [comment]
I can kind of sort of get it working from the 1.2.0 release branch/tag. There are still some issues:

1. autoscaling seems broken (despite ample cluster resources, the operator never adds more than `minWorkers` workers
2. port forwarding of the dashboard service fails with
```
E0323 16:17:39.345071   53588 portforward.go:400] an error occurred forwarding 8265 -> 8266: error forwarding port 8266 to pod 8821c7d6949d6150ae3ff0929bb39c58f7af1168d3f460dbfc62d221028ab855, uid : failed to execute portforward in network namespace "/var/run/netns/cni-5b8cb549-d51e-eec4-debe-7cbe4158a210": socat command returns error: exit status 1, stderr: "2021/03/23 20:17:39 socat[605203] E connect(5, AF=2 127.0.0.1:8266, 16): Connection refused\n"
```

**Comment by ssiegel95** [comment]
Any fix planned for[ the section](https://docs.ray.io/en/master/cluster/kubernetes.html#using-ray-client-to-connect-from-outside-the-kubernetes-cluster) that doesn't work?

**Comment by tbabej** [comment]
Turning on the debugging output, I can see the following:
```
INFO:ray.util.client.worker:Waiting for Ray to become ready on the server, retry in 5s...
INFO:ray.util.client.worker:Waiting for Ray to become ready on the server, retry in 10s...
INFO:ray.util.client.worker:Waiting for Ray to become ready on the server, retry in 15s...
```

Digging a little bit deeper, it looks like the culprit is not on the client side. Diving into the debugger, we can see that:

```python
>>> self.get_cluster_info(ray_client_pb2.ClusterInfoType.IS_INITIALIZED)                                                                                                                    
False
```

we're getting negative response over the gRPC channel to the initialization query.

**Comment by tbabej** [comment]
The problem is resolved for me if I update the client version to the current master (by building wheels from the repo). The situation could be summarized as the 1.2.0 ray client silently failing if interacting with current dev build of the ray cluster (the examples in the docs use nightly builds).

**Comment by ssiegel95** [comment]
@tbabej Thank you for the analysis, very helpful. I'll give the nightly client and image a try today.

**Comment by ssiegel95** [comment]
Success! Thanks @tbabej . Closing issue.

**Comment by ssiegel95** [comment]
One follow-up question. If I use a different namespace (deploy the operator into something other than namespace `ray`) I get the following messages from my port-forward when I try to run the exanple:

```
$  kubectl -n stus port-forward service/example-cluster-ray-head 10001:10001
Forwarding from 127.0.0.1:10001 -> 10001
Forwarding from [::1]:10001 -> 10001
Handling connection for 10001
E0401 15:16:01.812382    4567 portforward.go:400] an error occurred forwarding 10001 -> 10001: error forwarding port 10001 to pod 3d7ffd6c445a55b117f7ac107d6aae85724b1874c48534311923861dd26b3951, uid : failed to execute portforward in network namespace "/var/run/netns/cni-40dae698-393c-58a6-c86b-26c3918082b3": socat command returns error: exit status 1, stderr: "2021/04/01 19:16:01 socat[760334] E connect(5, AF=2 127.0.0.1:10001, 16): Connection refused\n"
Handling connection for 10001
```
namespace stus has the listening services:

```bash
$ kubectl -n stus get services
NAME                              TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                       AGE
example-cluster-ray-head          ClusterIP   172.21.226.125   <none>        10001/TCP,8265/TCP,8000/TCP   10m
```


Is there something magical about doing all this within a namespace named `ray`?

**Comment by tbabej** [comment]
Not a ray maintainer, but for what it's worth the namespaces have different network interfaces, so if the service is exposed through a different namespace, port forwarding does not work. That's pure K8s though, nothing to do with Ray AFAIK.

With respect to this issue however, I don't think we should close it. It is certainly unintended behaviour that the older version of the client (1.2.0) cannot connect to a newer server (what will become 1.3.0) and fails silently at that. At the very least the docs need to be updated to point this out, but hopefully the server is modified to be compatible with the older client. Hence we should reopen, and maybe rephrase the title to "Ray 1.2.0 fails to connect to Ray cluster on K8s (running master)"

**Comment by DmitriGekhtman** [comment]
Added a note to the docs in the above PR.
Current master version of Ray adds an error message when mismatched Ray versions are used.
